third_party/move/move-vm/runtime/src/interpreter.rs

```diff
@@ -98,33 +110,76 @@ impl Interpreter {
         extensions: &mut NativeContextExtensions,
         loader: &Loader,
     ) -> VMResult<Vec<Value>> {
-        Interpreter {
-            operand_stack: Stack::new(),
-            call_stack: CallStack::new(),
-            paranoid_type_checks: loader.vm_config().paranoid_type_checks,
-            access_control: AccessControlState::default(),
-            active_modules: HashSet::new(),
-        }
-        .execute_main(
-            loader,
+        InterpreterImpl::entrypoint(
+            function,
+            args,
             data_store,
             module_store,
             module_storage,
             gas_meter,
             traversal_context,
             extensions,
-            function,
-            args,
+            loader,
         )
     }
+}
+
+impl InterpreterImpl {
+    /// Entrypoint into the interpreter. All external calls need to be routed through this
+    /// function.
+    pub(crate) fn entrypoint(
+        function: LoadedFunction,
+        args: Vec<Value>,
+        data_store: &mut TransactionDataCache,
+        module_store: &LegacyModuleStorageAdapter,
+        module_storage: &impl ModuleStorage,
+        gas_meter: &mut impl GasMeter,
+        traversal_context: &mut TraversalContext,
+        extensions: &mut NativeContextExtensions,
+        loader: &Loader,
+    ) -> VMResult<Vec<Value>> {
+        let interpreter = InterpreterImpl {
+            operand_stack: Stack::new(),
+            call_stack: CallStack::new(),
+            paranoid_type_checks: loader.vm_config().paranoid_type_checks,
+            access_control: AccessControlState::default(),
+            active_modules: HashSet::new(),
+        };
+
+        if loader.vm_config().paranoid_type_checks {
+            interpreter.execute_main::<FullRuntimeTypeCheck>(
+                loader,
+                data_store,
+                module_store,
+                module_storage,
+                gas_meter,
+                traversal_context,
+                extensions,
+                function,
+                args,
+            )
+        } else {
+            interpreter.execute_main::<NullRuntimeTypeCheck>(
+                loader,
+                data_store,
+                module_store,
+                module_storage,
+                gas_meter,
+                traversal_context,
+                extensions,
+                function,
+                args,
+            )
+        }
+    }
 
     /// Main loop for the execution of a function.
     ///
     /// This function sets up a `Frame` and calls `execute_code_unit` to execute code of the
     /// function represented by the frame. Control comes back to this function on return or
     /// on call. When that happens the frame is changes to a new one (call) or to the one
     /// at the top of the stack (return). If the call stack is empty execution is completed.
-    fn execute_main(
+    fn execute_main<RTTCheck: RuntimeTypeCheck>(
```

Comment:
Love the generic bound name :)

Reply:
I meant this is better than `C: ...`

Reply:
Lol, don't know how a "confused" emoji ended up there - wanted to click on a smiley